README.md
# nautilus_save_tabs

Nautilus scripts to save/restore tabs

* SaveTabs saves URIs of tabs
* GetURIs is a sub script used by SaveTabs
* RestoreTabs restores saved tabs

## requirements
* nautilus
* ruby
* xdotool
* zenity

## Configuration

### save_tabs

`save_tabs` is a yaml file to configure the scripts.
if you need to use diffrent configuration from the default one,
`cp save_tabs.example save_tabs` and modify `save_tabs`

| variable | description | default value |
|:---------|:------------|:--------------|
| sleep_time_save | sleep time after each key input by xdotool (seconds) in SaveTabs | 0.03 |
| max_tab | maximum number of tabs to scan | 128 |
| min_tab | number of tabs to scan at the start of scanning | 4 |
| shortcut_geturis | shortcut key to invoke `GetURIs` script | "control+g" |
| sleep_time_restore | sleep time after each key input by xdotool (seconds) in RestoreTabs | 0.3 |
| select_uri_delay | delay between keystrokes to select uri in RestoreTabs (milliseconds) | 100 |
| save_tabs_path | file to save tabs | ENV['HOME'] + '/.local/share/nautilus/tabs' |
| socket_path | socket used by SaveTabs and GetURIs | '/tmp/nautilus_' + ENV['USER'] + '.socket' |

#### sleep and delay

increase `sleep_time_*` and/or `select_uri_delay` if the script is unstable.

### accels
modify scripts-accels to configure keyboard shortcuts for Nautilus >= 3.20

## Installation
`bundle install --path vendor/bundle` and `bundle exec rake install` will install everything.

## Usage
* to save tabs, invoke SaveTabs clicking right button of the mouse on the file or using
  keyboard shortcuts configured in the installation.
* to resotore tabs, invoke RestoreTabs when a tab is open. the current tab is
  replaced with the first one and the rest of tabs are inserted after the current tab.
* DO NOT touch mouse and keyboard while saving/restoring tabs.

## Development

scripts have been tested with Ubuntu 17.04 and Ruby 2.4.1.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/susurri/nautilus_save_tabs. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The software is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.1"></a>
## [1.0.1](https://github.com/susurri/nautilus_save_tabs/compare/v1.0.0...v1.0.1) (2017-08-19)


### Performance Improvements

* improve performance of selecting uri ([47138c6](https://github.com/susurri/nautilus_save_tabs/commit/47138c6))



<a name="1.0.0"></a>
# 1.0.0 (2017-07-31)


spec/array_spec.rb
require_relative '../scripts/SaveTabs'

RSpec.describe 'example' do
  let(:save_tabs) { Nautilus::SaveTabs.new }
  it 'unwrap' do
    property_of do
      len = integer(1..128)
      remain = integer(1..128)
      [len, remain, array(len) { string }]
    end.check(900) do |(len, remain, arr)|
      expect(save_tabs
          .unwrap_around(arr.take_wrapped_around(len + remain),
                         arr.take_wrapped_around(-(len + remain))))
        .to eql arr
    end
  end
end


spec/spec_helper.rb
require 'rantly/rspec_extensions' # for RSpec

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

  # The settings below are suggested to provide a good initial experience
  # with RSpec, but feel free to customize to your heart's content.
  #   # This allows you to limit a spec run to individual examples or groups
  #   # you care about by tagging them with `:focus` metadata. When nothing
  #   # is tagged with `:focus`, all examples get run. RSpec also provides
  #   # aliases for `it`, `describe`, and `context` that include `:focus`
  #   # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  #   config.filter_run_when_matching :focus
  #
  #   # Allows RSpec to persist some state between runs in order to support
  #   # the `--only-failures` and `--next-failure` CLI options. We recommend
  #   # you configure your source control system to ignore this file.
  #   config.example_status_persistence_file_path = "spec/examples.txt"
  #
  #   # Limits the available syntax to the non-monkey patched syntax that is
  #   # recommended. For more details, see:
  #   #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  #   config.disable_monkey_patching!
  #
  #   # This setting enables warnings. It's recommended, but in some cases may
  #   # be too noisy due to issues in dependencies.
  #   config.warnings = true
  #
  #   # Many RSpec users commonly either run the entire suite or an individual
  #   # file, and it's useful to allow more verbose output when running an
  #   # individual spec file.
  #   if config.files_to_run.one?
  #     # Use the documentation formatter for detailed output,
  #     # unless a formatter has already been configured
  #     # (e.g. via a command-line flag).
  #     config.default_formatter = "doc"
  #   end
  #
  #   # Print the 10 slowest examples and example groups at the
  #   # end of the spec run, to help surface which specs are running
  #   # particularly slow.
  #   config.profile_examples = 10
  #
  #   # Run specs in random order to surface order dependencies. If you find an
  #   # order dependency and want to debug it, you can fix the order by providing
  #   # the seed, which is printed after each run.
  #   #     --seed 1234
  #   config.order = :random
  #
  #   # Seed global randomization in this process using the `--seed` CLI option.
  #   # Setting this allows you to use `--seed` to deterministically reproduce
  #   # test failures related to randomization by passing the same `--seed` value
  #   # as the one that triggered the failure.
  #   Kernel.srand config.seed
end

class Array
  def take_wrapped_around(n)
    if n < 0
      (0..(-n) - 1).to_a.map { |i| reverse[i % size] }
    else
      (0..n - 1).to_a.map { |i| self[i % size] }
    end
  end
end


nautilus_save_tabs.rb
#!/usr/bin/env ruby

require 'socket'

# nautilus module
module Nautilus
  ## (1..30).select {|n| 60 % n == 0}
  #
  # array.each_slice
  # save tabs
  class SaveTabs
    SLEEP_TIME = 0.05
    MAX_TAB = 5
    KEY_DELAY = 1000

    SHORTCUT_GETURIS = 'control+g'.freeze
    SAVE_TABS_PATH = ENV['HOME'] + '.config/nautilus/save_tabs'
    SOCKET_PATH = '/tmp/nautilus_' + ENV['USER'] + '.socket'

    @uriss = ''

    def key(keys)
      p = 'xdotool key ' + keys
      #  p = "xdotool --delay" + KEY_DELAY + keys + "control+l control+c"
      system(p)
      sleep(SLEEP_TIME)
    end

    def read_socket
      Thread.start do
        Socket.unix_server_loop(SOCKET_PATH) do |sock, _addr|
          begin
            @uris = sock.read
            break
          rescue Errno::EPIPE => e
            p e
            break
          end
        end
      end
    end

    def uris
      thread = read_socket
      key(SHORTCUT_GETURIS)
      thread.join
      p @uris
      @uris
    end

    def first_tab
      key('alt+1')
      uris
    end

    def next_tab
      key('control+Page_Down')
      uris
    end

    def prev_tab
      key('control+Page_Up')
      uris
    end

    def nth_tab(n)
      n = [9, [0, n].max].min
      key("alt+#{n}")
      uris
    end

    def initialize
      uris_forward = [first_tab]
      MAX_TAB.times do
        uris_forward << next_tab
      end

      system("zenity --list \
                    --title='Session saved' \
                    --width=800 --height=600 \
                    --column='Name' --column='Value' \
                    'uris' #{uris_forward}")
    end
  end
end

Nautilus::SaveTabs.new


scripts/SaveTabs.rb
#!/usr/bin/env ruby

require 'socket'
require 'yaml'
require 'English'
require 'shellwords'
require 'cgi'

# Integer class
class Integer
  def divisors
    (1..self).select { |i| (self % i).zero? }
  end
end

# Array class
class Array
  def unrecur
    size.divisors.each do |i|
      u = each_slice(i).to_a.uniq
      return(u.first) if u.size == 1
    end
  end
end

# nautilus module
module Nautilus
  # configurations
  class Config
    attr_reader(:config)
    def initialize
      @config = {
        'sleep_time_save' => 0.03,
        'max_tab' => 128,
        'min_tab' => 4,
        'shortcut_geturis' => 'control+g',
        'save_tabs_path' => ENV['HOME'] + '/.local/share/nautilus/tabs',
        'socket_path' => '/tmp/nautilus_' + ENV['USER'] + '.socket'
      }
    end

    def load
      @config.merge!(
        YAML.load_file("#{ENV['HOME']}/.config/nautilus/save_tabs")
      )
    end
  end
  # save tabs
  class SaveTabs
    def key(keys)
      p = 'xdotool key ' + keys
      system(p)
      sleep(@config.config['sleep_time_save'])
    end

    def read_socket
      Thread.start do
        Socket
          .unix_server_loop(@config.config['socket_path']) do |sock, _addr|
          sock.puts('')
          @uris = YAML.safe_load(sock.read)
          break
        end
      end
    end

    def read_tab
      thread = read_socket
      key(@config.config['shortcut_geturis'])
      thread.join
      @uris
    end

    def first_tab
      key('alt+1')
      read_tab
    end

    def next_tab
      key('control+Page_Down')
      read_tab
    end

    def prev_tab
      key('control+Page_Up')
      read_tab
    end

    def last_tab
      first_tab
      prev_tab
    end

    def scan_forward(n)
      uris = [first_tab]
      (n - 1).times do
        uris << next_tab
      end
      uris
    end

    def scan_backward(n)
      uris = [last_tab]
      (n - 1).times do
        uris << prev_tab
      end
      uris
    end

    def head_tail_match?(x, i)
      x.first(i) == x.last(i)
    end

    def rotate_match?(x, y, i)
      j = y.size - i
      x.first(j) == y.rotate(i).first(j) && head_tail_match?(x, i) &&
        head_tail_match?(y, i)
    end

    # Array -> Array -> Array
    def unwrap_around(x, y)
      y.reverse!
      0.upto(y.size.div(2)) do |i|
        j = y.size - i
        return(x.first(j).unrecur) if rotate_match?(x, y, i)
      end
      []
    end

    def save_tabs(obj)
      File.open(@config.config['save_tabs_path'], 'w') do |f|
        f.write obj.to_yaml
      end
    end

    def numbers_of_tabs(min, max)
      0.upto(Math.log2(max / min)).to_a.map { |i| min * 2**i } + [max]
    end

    def warn_if_mismatch(x, y)
      command = 'zenity --warning --text "forward and backward scan mismatch"'
      system(command) if x.size != y.size
    end

    def scan_tabs
      numbers_of_tabs(@config.config['min_tab'],
                      @config.config['max_tab']).each do |n|
        uris_forward = scan_forward(n)
        uris_backward = scan_backward(n)
        warn_if_mismatch(uris_forward, uris_backward)
        l = unwrap_around(uris_forward, uris_backward)
        return(l) unless l.empty?
      end
      system('zenity --warning --text "max_tab is too small"')
      uris_forward
    end

    def uris_to_s(uris)
      str = ''
      uris.each do |u|
        str1 = if u[1].empty?
                 Shellwords.escape('-')
               else
                 Shellwords.escape(CGI.unescape(u[1].chomp.tr("\n", ','))
                          .encode(invalid: :replace))
               end
        str += (Shellwords.escape(CGI.unescape(u[0])) + ' ' + str1 + ' ')
      end
      str
    end

    def show_results(uris)
      command = "zenity --list --title='Tabs saved' --text ''\
            --width=800 --height=600 --column='URIs' --column='Selected URIs' "
      command += uris_to_s(uris)
      system(command)
    end

    def run
      @config = Config.new
      uris = scan_tabs
      save_tabs(uris)
      show_results(uris)
    end

    def initialize; end
  end
end

if $PROGRAM_NAME == __FILE__
  # sleep a second to ensure the key to invoke the script is released
  sleep(1)
  save_tabs = Nautilus::SaveTabs.new
  save_tabs.run
end
